face-detection-explainer.md

```diff
@@ -12,66 +12,125 @@
+};
+
+enum ObjectDetectionMode {
+  "none",          // Face or landmark detection is not needed
+  "center-point",  // Center point of the detected object is returned
+  "bounding-box",  // Bounding box of the detected object is returned
+};
+
+```
+
+## Using the face detection API
+
+### Metadata
+
+The first part of the API adds a new member `humanFaces` of type sequence of `HumanFace` into WebCodecs [`VideoFrameMetadata`](https://www.w3.org/TR/webcodecs/#dictdef-videoframemetadata) which provides information of the detected faces in the frame. In `HumanFace`, the member `id` is used to track faces between frames: the same `id` of a face between different frames indicates that it is the same face under tracking. However, it is specifically required that it must not be possible to correlate faces between different sources or video sequences by matching the `id` between them. If host uses face recognizion to track faces, it must assign a random integer to `id` between sequences to avoid privacy issues. `probability` is the probability that the returned face is in fact a human face and not a false detection.
+
+The member `boundingBox` provides the enclosing bounding box for the face.
+
+The members `leftEye`, `rightEye`, and `mouth` provide information on facial features belonging to the detected face. Currently this includes the center point for eyes and mouth, but we define a separate dictionary `HumanFaceLandmark` to allow extending it easily later on. 
+
+The coordinates in the members `boundingBox` and `centerPoint` of the detected faces and landmarks are defined similarly as in the
 [`pointsOfInterest`](https://w3c.github.io/mediacapture-image/#points-of-interest)
-field in [`MediaTrackSettings`](https://w3c.github.io/mediacapture-image/#dom-mediatracksettings-pointsofinterest)
-with the exception that the points may also lie outside of the frame since a detected face could be
-partially outside of the visible image.
-A `Point2D` is interpreted to represent a pixel position in a normalized square space. The origin of
-coordinates {x,y} = {0.0, 0.0} represents the upper leftmost corner whereas the {x,y} =
-{1.0, 1.0} represents the lower rightmost corner relative to the rendered frame: the x coordinate (columns) increases rightwards and the y
-coordinate (rows) increases downwards.
+member in [`MediaTrackSettings`](https://w3c.github.io/mediacapture-image/#dom-mediatracksettings-pointsofinterest)
+with the exception that the coordinates may also lie outside of the frame since a detected face could be
+partially (or in special cases even fully) outside of the visible image.
+A coordinate is interpreted to represent a location in a normalized square space. The origin of
+coordinates (x,y) = (0.0, 0.0) represents the upper leftmost corner whereas the (x,y) =
+(1.0, 1.0) represents the lower rightmost corner relative to the rendered frame: the x-coordinate (columns) increases rightwards and the y-coordinate (rows) increases downwards.

```

Comment:
Does it make sense to normalise?
DOMRectReadOnly has a width/height and an origin which would not be normalised but Point2D would be?
What is doing https://wicg.github.io/shape-detection-api? OS APIs?

Reply:
We had a discussion with the WebCodecs people (eg. Dan Sanders) and the issue of coordinate system turned to be a bit complicated. A normalized coordinate space seemed to be more straightforward and avoids some issues compared to natural size which doesn't always match with copyTo(). It is too bad that it doesn't seem to be consistent with the existing Web APIs. [Shape detection API](https://wicg.github.io/shape-detection-api) uses natural size (unlike we) but after the discussions it seems that the whole definition of coordinate space in that spec is not adequate, at least not for our purposes with VideoFrame which has internally a rotation, visibleRect, and codedRect. 

Therefore, I would rather stick with the definition from Image Capture spec's [Points of interest](https://www.w3.org/TR/image-capture/#points-of-interest) with normalized coordinates. I think that spec is more mainstream and avoids the above issues.

The platform APIs don't help here much, Windows uses normalized coordinates (up to INT_MAX or something) while Android/ChromeOS uses native pixel coordinates.

I think that both `Point2D` and `DOMRectReadOnly` types should use the same coordinate system for consistency. If we use normalized coordinates, also `width` and `height` should be normalized so that `right` = `x + width` for consistency. I will clarify that in the draft API once we agree on coordinate system.